### PR TITLE
Refactor qconv to reduce allocations.

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42840 Enable input pointer caching in XNNPACK integration.
* #42178 Change quantizer to account for input tensor's memory format.
* #42008 Call qnnpack's conv setup only if input pointer has changed.
* **#42007 Refactor qconv to reduce allocations.**
* #42006 Simple caching allocator for CPU.

Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc [hunk 1]
```diff
@@ -373,60 +350,23 @@ enum pytorch_qnnp_status qnnpackConv(
       const uint32_t cr = pytorch_qnnp_params.q8dw9.cr;
       const size_t group_stride = (groups + (cr - 1)) & -cr;
 
-      if (any_padding) {
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc:376]: Can you explain why this block is removed? From 376 to 414.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc [hunk 2]
```diff
@@ -173,45 +118,32 @@ enum pytorch_qnnp_status qnnpackDeConv(
...
-  deconvolution->output_width = output_width;
-  deconvolution->output = output;
-  deconvolution->output_pixel_stride = deconv_p.output_channels;
+  const size_t input_pixel_stride = deconv_p.input_channels;
+  const size_t output_pixel_stride = deconv_p.output_channels;
 
-  const void** indirection_buffer = (const void**)realloc(
-      deconvolution->indirection_buffer, indirection_buffer_size);
-  if (indirection_buffer == NULL) {
+  pytorch_qnnp_status status = pytorch_qnnp_setup_deconvolution2d_nhwc_q8(
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:132]: So again, all of the logic removed to the left is replaced with `pytorch_qnnp_setup_deconvolution2d_nhwc_q8`?

File: aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution-operator-tester.h [hunk 3]
```diff
@@ -24,6 +24,79 @@
...
+  convolution->dilation_width = conv_p.dilation[0];
+  convolution->input_padding_top = conv_p.padding[0];
+  convolution->input_padding_left = conv_p.padding[1];
+  convolution->input_padding_bottom = conv_p.padding[2];
+  convolution->input_padding_right = conv_p.padding[3];
+
+  const bool any_padding = (conv_p.padding[0]| conv_p.padding[1]
+      |conv_p.padding[2] | conv_p.padding[3]) != 0;
+
+  if (any_padding) {
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution-operator-tester.h:55]: OK I see, it is moved here.

**Review by AshkanAliabadi — APPROVED** [review]

File: aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h [hunk 4]
```diff
@@ -226,12 +226,6 @@ struct conv_param_t {
         ukernel_type = pytorch_qnnp_ukernel_type_conv;
       }
     }
-
-    if (per_channel && ukernel_type == pytorch_qnnp_ukernel_type_xzp_gemm) {
-      pytorch_qnnp_log_error(
-          "Per channel quantized weights are not supported for XZP kernels");
-      assert("Failed to initialize QNNPACK conv_param_t struct.");
-    }
```
[the remark this answers is not in the document]
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:234]: Thanks for the catch. Strange I removed this. Will reinstate.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc [hunk 5]
```diff
@@ -323,48 +324,24 @@ enum pytorch_qnnp_status qnnpackConv(
...
+  pytorch_qnnp_status status = pytorch_qnnp_setup_convolution2d_nhwc_q8(
+      convolution,
+      batch_size,
+      input_height,
+      input_width,
+      input,
+      input_pixel_stride,
+      output,
+      output_pixel_stride,
+      threadpool);
```
[the remark this answers is not in the document]
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc:337]: Yes.

**Pull request closed without merging**